File: modules/core/misc/java/test/RotatedRectTest.java
package org.opencv.test.core;

import org.opencv.core.CvType;
import org.opencv.core.Point;
import org.opencv.core.Rect;
import org.opencv.core.RotatedRect;
import org.opencv.core.MatOfRotatedRect;
import org.opencv.core.Size;
import org.opencv.test.OpenCVTestCase;

import java.util.Arrays;
import java.util.List;

public class RotatedRectTest extends OpenCVTestCase {

    private double angle;
    private Point center;
    private Size size;

    @Override
    protected void setUp() throws Exception {
        super.setUp();

        center = new Point(matSize / 2, matSize / 2);
        size = new Size(matSize / 4, matSize / 2);
        angle = 40;
    }

    public void testBoundingRect() {
        size = new Size(matSize / 2, matSize / 2);
        assertEquals(size.height, size.width);
        double length = size.height;

        angle = 45;
        RotatedRect rr = new RotatedRect(center, size, angle);

        Rect r = rr.boundingRect();
        double halfDiagonal = length * Math.sqrt(2) / 2;

        assertTrue((r.x == Math.floor(center.x - halfDiagonal)) && (r.y == Math.floor(center.y - halfDiagonal)));

        assertTrue((r.br().x >= Math.ceil(center.x + halfDiagonal)) && (r.br().y >= Math.ceil(center.y + halfDiagonal)));

        assertTrue((r.br().x - Math.ceil(center.x + halfDiagonal)) <= 1 && (r.br().y - Math.ceil(center.y + halfDiagonal)) <= 1);
    }

    public void testClone() {
        RotatedRect rrect = new RotatedRect(center, size, angle);
        RotatedRect clone = rrect.clone();

        assertTrue(clone != null);
        assertTrue(rrect.center.equals(clone.center));
        assertTrue(rrect.size.equals(clone.size));
        assertTrue(rrect.angle == clone.angle);
    }

    public void testEqualsObject() {
        Point center2 = new Point(matSize / 3, matSize / 1.5);
        Size size2 = new Size(matSize / 2, matSize / 4);
        double angle2 = 0;

        RotatedRect rrect1 = new RotatedRect(center, size, angle);
        RotatedRect rrect2 = new RotatedRect(center2, size2, angle2);
        RotatedRect rrect3 = rrect1;
        RotatedRect clone1 = rrect1.clone();
        RotatedRect clone2 = rrect2.clone();

        assertTrue(rrect1.equals(rrect3));
        assertTrue(!rrect1.equals(rrect2));

        assertTrue(rrect2.equals(clone2));
        clone2.angle = 10;
        assertTrue(!rrect2.equals(clone2));

        assertTrue(rrect1.equals(clone1));

        clone1.center.x += 1;
        assertTrue(!rrect1.equals(clone1));

        clone1.center.x -= 1;
        assertTrue(rrect1.equals(clone1));

        clone1.size.width += 1;
        assertTrue(!rrect1.equals(clone1));

        assertTrue(!rrect1.equals(size));
    }

    public void testHashCode() {
        RotatedRect rr = new RotatedRect(center, size, angle);
        assertEquals(rr.hashCode(), rr.hashCode());
    }

    public void testPoints() {
        RotatedRect rrect = new RotatedRect(center, size, angle);

        Point p[] = new Point[4];
        rrect.points(p);

        boolean is_p0_irrational = (100 * p[0].x != (int) (100 * p[0].x)) && (100 * p[0].y != (int) (100 * p[0].y));
        boolean is_p1_irrational = (100 * p[1].x != (int) (100 * p[1].x)) && (100 * p[1].y != (int) (100 * p[1].y));
        boolean is_p2_irrational = (100 * p[2].x != (int) (100 * p[2].x)) && (100 * p[2].y != (int) (100 * p[2].y));
        boolean is_p3_irrational = (100 * p[3].x != (int) (100 * p[3].x)) && (100 * p[3].y != (int) (100 * p[3].y));

        assertTrue(is_p0_irrational && is_p1_irrational && is_p2_irrational && is_p3_irrational);

        assertTrue("Symmetric points 0 and 2",
                Math.abs((p[0].x + p[2].x) / 2 - center.x) + Math.abs((p[0].y + p[2].y) / 2 - center.y) < EPS);

        assertTrue("Symmetric points 1 and 3",
                Math.abs((p[1].x + p[3].x) / 2 - center.x) + Math.abs((p[1].y + p[3].y) / 2 - center.y) < EPS);

        assertTrue("Orthogonal vectors 01 and 12",
                Math.abs((p[1].x - p[0].x) * (p[2].x - p[1].x) +
                        (p[1].y - p[0].y) * (p[2].y - p[1].y)) < EPS);

        assertTrue("Orthogonal vectors 12 and 23",
                Math.abs((p[2].x - p[1].x) * (p[3].x - p[2].x) +
                        (p[2].y - p[1].y) * (p[3].y - p[2].y)) < EPS);

        assertTrue("Orthogonal vectors 23 and 30",
                Math.abs((p[3].x - p[2].x) * (p[0].x - p[3].x) +
                        (p[3].y - p[2].y) * (p[0].y - p[3].y)) < EPS);

        assertTrue("Orthogonal vectors 30 and 01",
                Math.abs((p[0].x - p[3].x) * (p[1].x - p[0].x) +
                        (p[0].y - p[3].y) * (p[1].y - p[0].y)) < EPS);

        assertTrue("Length of the vector 01",
                Math.abs((p[1].x - p[0].x) * (p[1].x - p[0].x) +
                        (p[1].y - p[0].y) * (p[1].y - p[0].y) - size.height * size.height) < EPS);

        assertTrue("Length of the vector 21",
                Math.abs((p[1].x - p[2].x) * (p[1].x - p[2].x) +
                        (p[1].y - p[2].y) * (p[1].y - p[2].y) - size.width * size.width) < EPS);

        assertTrue("Angle of the vector 21 with the axes", Math.abs((p[2].x - p[1].x) / size.width - Math.cos(angle * Math.PI / 180)) < EPS);
    }

    public void testRotatedRect() {
        RotatedRect rr = new RotatedRect();

        assertTrue(rr != null);
        assertTrue(rr.center != null);
        assertTrue(rr.size != null);
        assertTrue(rr.angle == 0.0);
    }

    public void testRotatedRectDoubleArray() {
        double[] vals = { 1.5, 2.6, 3.7, 4.2, 5.1 };
        RotatedRect rr = new RotatedRect(vals);

        assertNotNull(rr);
        assertEquals(1.5, rr.center.x);
        assertEquals(2.6, rr.center.y);
        assertEquals(3.7, rr.size.width);
        assertEquals(4.2, rr.size.height);
        assertEquals(5.1, rr.angle);
    }

    public void testRotatedRectPointSizeDouble() {
        RotatedRect rr = new RotatedRect(center, size, 40);

        assertTrue(rr != null);
        assertTrue(rr.center != null);
        assertTrue(rr.size != null);
        assertTrue(rr.angle == 40.0);
    }

    public void testSet() {
        double[] vals1 = {};
        RotatedRect r1 = new RotatedRect(center, size, 40);

        r1.set(vals1);

        assertEquals(0., r1.angle);
        assertPointEquals(new Point(0, 0), r1.center, EPS);
        assertSizeEquals(new Size(0, 0), r1.size, EPS);

        double[] vals2 = { 1, 2, 3, 4, 5 };
        RotatedRect r2 = new RotatedRect(center, size, 40);

        r2.set(vals2);

        assertEquals(5., r2.angle);
        assertPointEquals(new Point(1, 2), r2.center, EPS);
        assertSizeEquals(new Size(3, 4), r2.size, EPS);
    }

    public void testToString() {
        String actual = new RotatedRect(new Point(1, 2), new Size(10, 12), 4.5).toString();
        String expected = "{ {1.0, 2.0} 10x12 * 4.5 }";
        assertEquals(expected, actual);
    }

    public void testMatOfRotatedRect() {
        RotatedRect a = new RotatedRect(new Point(1,2),new Size(3,4),5.678);
        RotatedRect b = new RotatedRect(new Point(9,8),new Size(7,6),5.432);
        MatOfRotatedRect m = new MatOfRotatedRect(a,b,a,b,a,b,a,b);
        assertEquals(m.rows(), 8);
        assertEquals(m.cols(), 1);
        assertEquals(m.type(), CvType.CV_32FC(5));
        RotatedRect[] arr = m.toArray();
        assertEquals(arr[2].angle, a.angle, EPS);
        assertEquals(arr[3].center.x, b.center.x);
        assertEquals(arr[3].size.width, b.size.width);
        List<RotatedRect> li = m.toList();
        assertEquals(li.size(), 8);
        RotatedRect rr = li.get(7);
        assertEquals(rr.angle, b.angle, EPS);
        assertEquals(rr.center.y, b.center.y);
    }
}

File: modules/core/misc/objc/test/swift/RotatedRectTest.swift
//
//  RotatedRectTest.swift
//
//  Created by Giles Payne on 2020/01/31.
//

import XCTest
import OpenCV

class RotatedRectTest: OpenCVTestCase {

    let angle: Double = 40
    let center = Point2f(x: Float(OpenCVTestCase.matSize / 2), y: Float(OpenCVTestCase.matSize / 2))
    let size = Size2f(width: Float(OpenCVTestCase.matSize / 4), height: Float(OpenCVTestCase.matSize / 2))

    func testBoundingRect() {
        let size = Size2f(width: Float(OpenCVTestCase.matSize / 2), height: Float(OpenCVTestCase.matSize / 2));
        XCTAssertEqual(size.height, size.width);
        let length = size.height;

        let angle: Double = 45
        let rr = RotatedRect(center: center, size: size, angle: angle)

        let r = rr.boundingRect()
        let halfDiagonal = length * sqrt(2) / 2

        XCTAssert(Float(r.x) == floor(center.x - halfDiagonal) && Float(r.y) == floor(center.y - halfDiagonal))

        XCTAssert((r.br().x >= ceil(center.x + halfDiagonal)) && (r.br().y >= ceil(center.y + halfDiagonal)))

        XCTAssert((r.br().x - ceil(center.x + halfDiagonal)) <= 1 && (r.br().y - ceil(center.y + halfDiagonal)) <= 1)
    }

    func testClone() {
        let rrect = RotatedRect(center: center, size: size, angle: angle)
        let clone = rrect.clone();

        XCTAssertNotNil(clone)
        XCTAssert(rrect.center == clone.center)
        XCTAssert(rrect.size == clone.size)
        XCTAssert(rrect.angle == clone.angle)
    }

    func testEqualsObject() {
        let center2 = Point2f(x: Float(OpenCVTestCase.matSize / 3), y: Float(OpenCVTestCase.matSize) / 1.5)
        let size2 = Size2f(width: Float(OpenCVTestCase.matSize / 2), height: Float(OpenCVTestCase.matSize / 4))
        let angle2:Double = 0

        let rrect1 = RotatedRect(center: center, size: size, angle: angle)
        let rrect2 = RotatedRect(center: center2, size: size2, angle: angle2)
        let rrect3 = rrect1
        let clone1 = rrect1.clone()
        let clone2 = rrect2.clone()

        XCTAssert(rrect1 == rrect3)
        XCTAssertFalse(rrect1 == rrect2)

        XCTAssert(rrect2 == clone2)
        clone2.angle = 10
        XCTAssertFalse(rrect2 == clone2)

        XCTAssert(rrect1 == clone1)

        clone1.center.x += 1
        XCTAssertFalse(rrect1 == clone1)

        clone1.center.x -= 1
        XCTAssert(rrect1 == clone1)

        clone1.size.width += 1
        XCTAssertFalse(rrect1 == clone1)

        XCTAssertFalse(rrect1 == size)
    }

    func testHashCode() {
        let rr = RotatedRect(center: center, size: size, angle: angle)
        XCTAssertEqual(rr.hash(), rr.hash())
    }

    func testPoints() {
        let rrect = RotatedRect(center: center, size: size, angle: angle);

        let p = rrect.points()

        let is_p0_irrational = (100 * p[0].x != round(100 * p[0].x)) && (100 * p[0].y != round(100 * p[0].y))
        let is_p1_irrational = (100 * p[1].x != round(100 * p[1].x)) && (100 * p[1].y != round(100 * p[1].y));
        let is_p2_irrational = (100 * p[2].x != round(100 * p[2].x)) && (100 * p[2].y != round(100 * p[2].y));
        let is_p3_irrational = (100 * p[3].x != round(100 * p[3].x)) && (100 * p[3].y != round(100 * p[3].y));

        XCTAssert(is_p0_irrational && is_p1_irrational && is_p2_irrational && is_p3_irrational)

        XCTAssert(abs((p[0].x + p[2].x) / 2 - center.x) + abs((p[0].y + p[2].y) / 2 - center.y) < OpenCVTestCase.FEPS, "Symmetric points 0 and 2")

        XCTAssert(abs((p[1].x + p[3].x) / 2 - center.x) + abs((p[1].y + p[3].y) / 2 - center.y) < OpenCVTestCase.FEPS, "Symmetric points 1 and 3")

        XCTAssert(abs((p[1].x - p[0].x) * (p[2].x - p[1].x) +
            (p[1].y - p[0].y) * (p[2].y - p[1].y)) < OpenCVTestCase.FEPS, "Orthogonal vectors 01 and 12")

        XCTAssert(abs((p[2].x - p[1].x) * (p[3].x - p[2].x) +
            (p[2].y - p[1].y) * (p[3].y - p[2].y)) < OpenCVTestCase.FEPS, "Orthogonal vectors 12 and 23");

        XCTAssert(abs((p[3].x - p[2].x) * (p[0].x - p[3].x) +
            (p[3].y - p[2].y) * (p[0].y - p[3].y)) < OpenCVTestCase.FEPS, "Orthogonal vectors 23 and 30")

        XCTAssert(abs((p[0].x - p[3].x) * (p[1].x - p[0].x) +
            (p[0].y - p[3].y) * (p[1].y - p[0].y)) < OpenCVTestCase.FEPS, "Orthogonal vectors 30 and 01")

        XCTAssert(abs((p[1].x - p[0].x) * (p[1].x - p[0].x) +
            (p[1].y - p[0].y) * (p[1].y - p[0].y) - size.height * size.height) < OpenCVTestCase.FEPS, "Length of the vector 01")

        XCTAssert(abs((p[1].x - p[2].x) * (p[1].x - p[2].x) +
            (p[1].y - p[2].y) * (p[1].y - p[2].y) - size.width * size.width) < OpenCVTestCase.FEPS, "Length of the vector 21")

        XCTAssert(abs((p[2].x - p[1].x) / size.width - Float(cos(angle * Double.pi / 180))) < OpenCVTestCase.FEPS, "Angle of the vector 21 with the axes");
    }

    func testRotatedRect() {
        let rr = RotatedRect()

        XCTAssertNotNil(rr)
        XCTAssertNotNil(rr.center)
        XCTAssertNotNil(rr.size)
        XCTAssertEqual(0.0, rr.angle)
    }

    func testRotatedRectDoubleArray() {
        let vals = [1.5, 2.6, 3.7, 4.2, 5.1]
        let rr = RotatedRect(vals: vals as [NSNumber])

        XCTAssertNotNil(rr)
        XCTAssertEqual(1.5, rr.center.x)
        XCTAssertEqual(2.6, rr.center.y)
        XCTAssertEqual(3.7, rr.size.width)
        XCTAssertEqual(4.2, rr.size.height)
        XCTAssertEqual(5.1, rr.angle)
    }

    func testRotatedRectPointSizeDouble() {
        let rr = RotatedRect(center: center, size: size, angle: 40);

        XCTAssertNotNil(rr)
        XCTAssertNotNil(rr.center)
        XCTAssertNotNil(rr.size)
        XCTAssertEqual(40.0, rr.angle);
    }

    func testSet() {
        let vals1: [Double] = []
        let r1 = RotatedRect(center: center, size: size, angle: 40);

        r1.set(vals: vals1 as [NSNumber])

        XCTAssertEqual(0, r1.angle)
        assertPoint2fEquals(Point2f(x: 0, y: 0), r1.center, OpenCVTestCase.FEPS)
        assertSize2fEquals(Size2f(width: 0, height: 0), r1.size, OpenCVTestCase.FEPS)

        let vals2 = [1, 2, 3, 4, 5]
        let r2 = RotatedRect(center: center, size: size, angle: 40)

        r2.set(vals: vals2 as [NSNumber])

        XCTAssertEqual(5, r2.angle)
        assertPoint2fEquals(Point2f(x: 1, y: 2), r2.center, OpenCVTestCase.FEPS)
        assertSize2fEquals(Size2f(width: 3, height: 4), r2.size, OpenCVTestCase.FEPS)
    }

    func testToString() {
        let actual = "\(RotatedRect(center: Point2f(x:1, y:2), size: Size2f(width:10, height:12), angle:4.5))"
        let expected = "RotatedRect {Point2f {1.000000,2.000000},Size2f {10.000000,12.000000},4.500000}"
        XCTAssertEqual(expected, actual);
    }

}
